llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -26258,6 +26258,24 @@ SDValue DAGCombiner::visitFP_TO_FP16(SDNode *N) {
   if (N0->getOpcode() == ISD::FP16_TO_FP)
     return N0->getOperand(0);
 
+  // fold (fp_to_fp16 (freeze (fp16_to_fp (fp_to_fp16 op))))
```

Comment:
(Commenting in the right place this time)

This is a specific instance of isGuaranteedNotToBePoison; it would be better to reframe this in terms of a new DAG clone of the ValueTracking helper

Reply:
> I'm slightly abusing the lack of freeze optimisations to use freeze as a general optimisation barrier.

That does not sound safe to me.

Reply:
> That does not sound safe to me.

Understandable. It's still safer than the status quo where we miscompile code and have tests to ensure that we miscompile code, but if we can do better that would be great.

The two options I see are adding that dummy node type and then treating that exactly as I am currently treating `FREEZE`, or just saying that `FP_ROUND` and `FP_EXTEND` do not canonicalise and ignoring any performance degradations due to redundant recanonicalisations. Would either of those approaches be acceptable? Or is there some other option that I am just not seeing?

Reply:
> > That does not sound safe to me.
> 
> Understandable.

That was a bit of a drive-by comment (sorry) since I am not deeply involved in this code. I'm not sure I understand the underlying problem. Why is it OK to optimize away a rounding, if we are relying on it also canonicalizing? Is this because we're running in a mode where general fp operations are not required to canonicalize, but we do want to preserve explicit canonicalization ops?

Reply:
Yes, exactly that. Aside from SNaN handling, `(fp_round (fp_extend x))` and `(fp_extend (fp_round x, /*trunc*/1))` are always equal to `x` and so for code that does not care about SNaNs, these are useful to optimise away. But `(fp_round (fcanonicalize (fp_extend x)))` and `(fp_extend (fcanonicalize (fp_round x, /*trunc*/1)))` must handle SNaNs, so we cannot just optimise the `fcanonicalize` to a no-op on the basis that `fp_extend`/`fp_round` will already have canonicalized.

Reply:
Off the top of my head we could either:
1. Add a "must canonicalize" version of fp_round and/or fp_extend. But this does not sound very extendable, since we might also need "must canonicalize" versions of any other fp op that could be optimized away. Or:
2. Do not optimize away fcanonicalize as a DAG combine, but detect at instruction selection time when it is redundant because the input is already canonicalized. I guess this is what @arsenm suggested earlier (sorry I'm just catching up now). Is it hard to do this (without significant regressions)?

Reply:
1) indeed does not sound very extendable, and I suspect we will get more optimisable operations in the future.
2) is what @arsenm briefly suggested earlier, before suggesting that maybe it should be done even later. It's very hard to do without regressions, but whether any of the regressions will count as *significant* is hard to say in advance.

Reply:
Let's just bite the bullet and do it, I guess. Some of the regressions are easily solved and I have included them in the same commit. Others are not but the degradations that remain actually seem like they are outweighed by better `fcanonicalize` optimisations elsewhere.